polar-core/src/kb.rs

```diff
@@ -193,8 +224,12 @@ impl KnowledgeBase {
         rule_type_instance: &InstanceLiteral,
         index: usize,
     ) -> PolarResult<RuleParamMatch> {
+        // TODO @patrickod remove this hacky comparison edge-case by making Actor and Resource "real" classes
+        if (rule_instance.tag.0 == RESOURCE_UNION_NAME) || (rule_instance.tag.0 == ACTOR_UNION_NAME)
```

Comment:
I'm not sure this logic is correct.

At this call site, `rule_instance` isn't a union b/c we would've fallen into the case on 281 and `rule_type_instance` isn't a union b/c a union can't currently be a member of another union, and we've constructed `rule_type_instance` from the members of the `rule_type_instance` union value that caused us to enter this `else if` block:

https://github.com/osohq/oso/blob/62ff5aeb85e4cacb800e64faea6fea2e7c7a4cfb/polar-core/src/kb.rs#L311

So that call site won't ever hit this `if` case.

At the other call site, `rule_instance` *might* be a union and `rule_type_instance` will never be a union (b/c it would've fallen into the branch on 280):

https://github.com/osohq/oso/blob/62ff5aeb85e4cacb800e64faea6fea2e7c7a4cfb/polar-core/src/kb.rs#L337

So if `rule_instance` is a union here, which will hit this `if` case, then `rule_type_instance` *won't* be a union, and unions are less specific than members of unions, so it should be a `RuleParamMatch::False`, no?